### PR TITLE
Add startAt/endAt cursors and limitToLast to MockQuery

### PR DESCRIPTION
**Pull request by maxluchterhand1**
Hey,

I needed the functionality of the Query.startAt/Query.endAt methods for some tests in my project.
I implemented them in a way which only simulates the cursor on a concrete match, that means

[0, 1, 2, 3, 4, 5, 7, 8, 9].startAt(5) = [5,7,8,9]

[0, 1, 2, 3, 4, 5, 7, 8, 9].startAt(6) = [0, 1, 2, 3, 4, 5, 7, 8, 9]

The behavior of endAt currently slightly differs from the actual Firestore behavior.
 
In this implementation, endAt makes the query end at the **first** value that matches the endAt value. In Firestore, it ends on the **last** value. This is gonna be addressed in #166.

Also, this implementation only supports exact matches. This is going to be addressed in #167.

### CODE REVIEW

**Comment by atn832** [comment]
Thanks for your PR @maxluchterhand1. I'll take a look this weekend. In the meantime, would you mind writing unit tests for your new code?

File: lib/src/mock_query.dart [hunk 1]
```diff
@@ -122,6 +128,37 @@ class MockQuery extends Mock implements Query {
...
 
+  @override
+  Query startAt(List<dynamic> values) =>
+      _cursorUtil(values, (docs, index) => docs.sublist(max(0, index)));
+
+  @override
+  Query endAt(List<dynamic> values) => _cursorUtil(values,
+      (docs, index) => docs.sublist(0, index == -1 ? docs.length : index));
+
+  Query _cursorUtil(
```
Comment (atn832) [inline lib/src/mock_query.dart:139]: Would you mind documenting this function, in particular what the f argument is supposed to be?

File: lib/src/mock_query.dart [hunk 2]
```diff
@@ -122,6 +128,37 @@ class MockQuery extends Mock implements Query {
...
+        'You can only specify as many start values as there are orderBy filters.',
+      );
+      if (docs.isEmpty) return docs;
+
+      var res = List.of(docs);
+      for (var i = 0; i < values.length; i++) {
+        final index = docs.indexWhere(
+          (doc) => doc.data()?[parameters['orderedBy'][i]] == values[i],
+        );
+        res = f(docs, index);
```
Comment (atn832) [inline lib/src/mock_query.dart:156]: Should it be `f(docs,...` or `f(res,...`?

File: lib/src/mock_query.dart (hunk 1 above)
Comment (atn832) [inline lib/src/mock_query.dart:139]: It'd help make the code slightly more readable if you passed `List<String> orderByKeys` as a parameter to this function. Would you mind doing that, so that _cursorUtil doesn't access parameters directly?

File: lib/src/mock_query.dart (hunk 2 above)
Comment (maxluchterhand1) [reply lib/src/mock_query.dart:156]: Yea, I think it's supposed to be f(res,...) actually.

**Comment by atn832** [comment]
I tried your code with the following tests:

```dart
  test('startAt, endAt', () async {
    final instance = MockFirestoreInstance();
    await instance.collection('temperatures').add({'temperature': 3});
    await instance.collection('temperatures').add({'temperature': 5});
    await instance.collection('temperatures').add({'temperature': 7});
    await instance.collection('temperatures').add({'temperature': 2});

    // Prints out (2, 3, 5, 7).
    print(
        (await instance.collection('temperatures').orderBy('temperature').get())
            .docs
            .map((s) => s.get('temperature')));

    final aboveFive = await instance
        .collection('temperatures')
        .orderBy('temperature')
        .startAt([5]).get();
    expect(aboveFive.docs.map((s) => s.get('temperature')), equals([5, 7]));

    final aboveZero = await instance
        .collection('temperatures')
        .orderBy('temperature')
        .startAt([0]).get();
    expect(
        aboveZero.docs.map((s) => s.get('temperature')), equals([2, 3, 5, 7]));

    final underSix = await instance
        .collection('temperatures')
        .orderBy('temperature')
        .endAt([6]).get();
    // Also prints out (2, 3, 5, 7).
    print(underSix.docs.map((s) => s.get('temperature')));
    expect(underSix.docs.map((s) => s.get('temperature')), [2, 3, 5]);

    final underOne = await instance
        .collection('temperatures')
        .orderBy('temperature')
        .endAt([1]).get();
    // Also prints out (2, 3, 5, 7).
    print(underOne.docs.map((s) => s.get('temperature')));
    expect(underOne.docs.isEmpty, isTrue);
  });
```

startAt works but not endsAt. For example underSix throws:
> Expected: [2, 3, 5]
>  Actual: MappedListIterable<MockQueryDocumentSnapshot, dynamic>:[2, 3, 5, 7]
>   Which: at location [3] is MappedListIterable<MockQueryDocumentSnapshot, dynamic>:[2, 3, 5, 7] which longer than expected

 Would you mind adding these tests and a few more, perhaps with two properties, to make sure the code works fine?

File: test/mock_query_test.dart [hunk 3]
```diff
@@ -825,4 +825,76 @@ void main() {
...
+
+    expect(snapshots.docs, hasLength(3));
+
+    snapshots = await baseQuery.startAt(['Springfield', 'Missouri']).get();
+
+    expect(snapshots.docs, hasLength(2));
+
+    snapshots = await baseQuery.endAt(['Springfield']).get();
+
+    expect(snapshots.docs, hasLength(2));
```
Comment (atn832) [inline test/mock_query_test.dart:889]: This doesn't match the actual Firestore. I got `{name: Los Angeles, state: California}, {name: Springfield, state: Massachusetts}, {name: Springfield, state: Missouri}, {name: Springfield, state: Wisconsin}` instead. It seems like they ended it at the last document that still matches the endAt value.

File: test/mock_query_test.dart [hunk 4]
```diff
@@ -825,4 +825,76 @@ void main() {
...
+
+    // Since there is no Springfield, Florida in our docs, it should ignore the second orderBy value
+    snapshots = await baseQuery.endAt(['Springfield', 'Florida']).get();
+
+    expect(snapshots.docs, hasLength(2));
```
Comment (atn832) [inline test/mock_query_test.dart:894]: This doesn't match the actual Firestore. I got `{name: Los Angeles, state: California}`. It seems like it stopped before hitting {name: Springfield, state: Missouri}, because Florida < Missouri.

File: test/mock_query_test.dart (hunk 3 above)
Comment (atn832) [reply test/mock_query_test.dart:889]: To prevent this PR from getting too complicated, I filed #166. Can you add a note here to explain it is different from Firestore, and that it is tracked at #166? We can tackle this edge case later.

File: test/mock_query_test.dart (hunk 4 above)
Comment (atn832) [reply test/mock_query_test.dart:894]: As you reminded me last time, this is expected as per your implementation, so I filed #167. Can you leave a note here to mention it's different from Firestore, and that it will be fixed by #167? This way whoever resolves the ticket later will know this unit test should be fixed as well.

File: test/mock_query_test.dart [hunk 5]
```diff
@@ -825,4 +825,76 @@ void main() {
...
+        instance.collection('cities').orderBy('name').orderBy('state');
+
+    var snapshots = await baseQuery.startAt(['Springfield']).get();
+
+    expect(snapshots.docs, hasLength(3));
+
+    // Since there is no Springfield, Florida in our docs, it should ignore the second orderBy value
+    snapshots = await baseQuery.startAt(['Springfield', 'Florida']).get();
+
+    expect(snapshots.docs, hasLength(3));
```
[the remark this answers is not in the document]
Comment (atn832) [reply test/mock_query_test.dart:881]: Although it's by chance since Floria happens to be alphabetically smaller than those three state names, so Firestore returns the same thing.

**Review by atn832 — APPROVED** [review]
Thanks for the detailed explanation and thorough unit tests. I only have minor comments about the tests. Would you mind resolving them? Then I'll merge your PR and publish an update.

**Comment by maxluchterhand1** [comment]
Hey, thanks for your feedback. I just took care of everything you asked for.

File: test/mock_query_test.dart [hunk 6]
```diff
@@ -8,6 +8,11 @@ import 'query_snapshot_matcher.dart';
 
 const uid = 'abc';
 
+extension ToData on List<QueryDocumentSnapshot> {
```
Comment (atn832) [inline test/mock_query_test.dart:11]: Beautiful

**Comment by atn832** [comment]
Thanks again for the PR. I published it to https://pub.dev/packages/cloud_firestore_mocks/changelog#083.

Also I annotated some of your tests when they were different with Firestore for future reference at https://github.com/atn832/cloud_firestore_mocks/commit/39f0ae41ee7842bbf613d5f5eee1b826ad587f9d.
